### ISSUE TITLE
Rework the user filter

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/AdguardForAndroid#1279]
This all started here:
https://github.com/AdguardTeam/AdguardForAndroid/issues/1255

However, I am not happy with how it works at the moment, so let's rework it properly.

### Toolbar menu

![adguard for android - user filter moqups 2017-06-13 12-48-23](https://user-images.githubusercontent.com/5947035/27076697-a4cb701e-5036-11e7-875e-777cac0ae6b9.png)

* Add rule -- brings up a "New Rule" dialog
* Remove all -- removes all rules from the user filter. **IMPORTANT:** we must show a confirmation dialog before actually doing it.
* Edit filter -- switches user filter to the "edit mode" (multiline text edit)
* Import -- brings up the "Import" dialog
* Export -- brings up the "Export" dialog

### Empty user filter

Screenshot:
<details>

![adguard for android - user filter moqups 2017-06-13 12-47-11](https://user-images.githubusercontent.com/5947035/27076645-7ed032e6-5036-11e7-8ef0-fe3fee587ba6.png)
</details>

### New Rule dialog

![adguard for android - user filter moqups 2017-06-13 12-51-21](https://user-images.githubusercontent.com/5947035/27076877-0db38bac-5037-11e7-881e-0f9560abb0d8.png)

**Important notes:**

* Use multiline text input (max rows == 3)
* Text input font should be `monospace`

### User filter rules list view

Screenshot:
<details>

![adguard for android - user filter moqups 2017-06-13 12-53-28](https://user-images.githubusercontent.com/5947035/27076971-53f386da-5037-11e7-8523-b12ed18d9e9e.png)
</details>

**Important notes:**

* Use monospace font for list view items
* Do not allow word-wrapping
* Depending on the rule text we should use different color for the rules text

Here are the rules for syntax highlighting:
```
// Depending on the substring we choose class
						{regex: /\!.*/, token: "comment"},
						{regex: /@@.*/,  token: "whitelist"},
						{regex: /%%.*/, token: "js"},
						{regex: /#%#.*/, token: "js"},
						{regex: /#@%#.*/, token: "js"},
						{regex: /#\$#.*/, token: "css-inject"},
						{regex: /#@\$#.*/, token: "css-inject"},
						{regex: /##.*/, token: "css"},
						{regex: /#@#.*/, token: "css"},
						{regex: /\$\$.*/, token: "content"},
						{regex: /\$@\$.*/, token: "content"}

// Rules colors
.cm-s-default .cm-comment {color: #808080;}
.cm-s-default .cm-whitelist {color: #336666;}
.cm-s-default .cm-css {color: #6699cc;}
.cm-s-default .cm-js {color: #999933;}
.cm-s-default .cm-css-inject {color: #4477aa;}
.cm-s-default .cm-content {color: #a176a2;}
```

### Edit rule dialog

Tapping on a list item brings up the "Edit Rule" dialog:
![adguard for android - user filter moqups 2017-06-13 12-57-02](https://user-images.githubusercontent.com/5947035/27077173-df221d66-5037-11e7-85ae-5c7a7a71fdf1.png)

It is almost the same as the "New Rule" save for one more option -- "Delete".
**IMPORTANT:** there should be a confirmation dialog after the user taps "delete".

### User filter edit mode

Choosing "Edit filter" in the context menu leads a user to the "user filter edit mode".

Here is how it looks like:
<details>

![adguard for android - user filter moqups 2017-06-13 12-59-06](https://user-images.githubusercontent.com/5947035/27077275-2a9a17e4-5038-11e7-962b-669fb10c81b9.png)
</details>

Basically, this is just a huge multiline text input with "monospace" font and two options in the context menu: `Save` and `Cancel`

**Comment by TPS** [comment]
#1210 now dupes this, too, though I'd to remind y'all of [my request from there](https://github.com/AdguardTeam/AdguardForAndroid/issues/1210#issuecomment-304875893):
>When in filter-editing mode (either from filtering-log or anyplace else), it'd be great to have additional toggles like `$important` for `$image`, `$script`, &c.

Also, #828, please.

**Comment by DrDidi** [comment]
Pls retain the rules order when ex-/importing (unlike today: https://forum.adguard.com/index.php?threads/bug-user-filter-import-puts-rules-in-random-order.22561/)

**Comment by ameshkov** [comment]
> Pls retain the rules order when ex-/importing

Sure, it's done in that alpha build you grabbed in the telegram channel.

**Comment by ameshkov** [comment]
> When in filter-editing mode (either from filtering-log or anyplace else), it'd be great to have additional toggles like $important for $image, $script, &c.

It will be simply too much for this task. I mean just look at how many different modifiers are out there:
https://kb.adguard.com/en/general/how-to-create-your-own-ad-filters

It should be filed as a separate feature request for sure.

> Also, #828, please.

Does "User filter edit mode" already cover it?

**Comment by ameshkov** [comment]
Reworked user filter:
[adguard_2.9.124_dev.apk.zip](https://github.com/AdguardTeam/AdguardForAndroid/files/1096167/adguard_2.9.124_dev.apk.zip)

@vozersky I need you to test it very thoroughly.

What to check:

1. User filter adding new rule
2. Removing existing rule
3. Editing existing rule
4. Import from a valid file
5. Import from an invalid file
6. Search in the user filter
7. Removing all the rules
8. Enabling/disabling a rule

Same for the whitelist.

**NOTE:** Don't forget to check that the changes are actually applied to filtering.

@Revertron the only issue I am aware of at the moment is the problem with action bar icons when search starts: https://monosnap.com/file/gC3u7EXFE4OpnjhcOxO4drcKl3Yo7W

**Comment by vozersky** [comment]
@ameshkov 

everything works correctly except point 5: **Import from an invalid file** 

**STR**
1. Try to import image file as a filter list.

**Expected result**
Toast Error Message is shown

**Actual result**

<details><summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/17472907/27588150-d17d73d6-5b4f-11e7-8e66-4cb6d206974c.png)

</details>

**Comment by ameshkov** [comment]
Should be re-tested when release candidate version is out today

**Comment by ameshkov** [comment]
@vozersky please add it to the regular Android test cases.

**Comment by vozersky** [comment]
Added

**Comment by TPS** [comment]
@AdGuardTeam Really excellent work!  Thanks!

**Comment by ilvjyw** [comment]
Hope to add：
Keep the original rule when importing

**Comment by vozersky** [comment]
@ilvjyw https://github.com/AdguardTeam/AdguardForAndroid/issues/1312
